synapse_token_authenticator/token_authenticator.py

```diff
@@ -145,37 +183,89 @@ async def check_auth(
         logger.info("All done and valid, logging in!")
         return (user_id_str, None)
 
+    async def check_oidc_auth(
+        self, username: str, login_type: str, login_dict: "synapse.module_api.JsonDict"
+    ) -> Optional[
+        Tuple[
+            str,
+            Optional[Callable[["synapse.module_api.LoginResponse"], Awaitable[None]]],
+        ]
+    ]:
+        logger.info("Receiving auth request")
+        if login_type != "com.famedly.login.token.oidc":
+            logger.info("Wrong login type")
+            return None
+        if "token" not in login_dict:
+            logger.info("Missing token")
+            return None
+        token = login_dict["token"]
+
+        oidc = self.config.oidc
+        oidc_metadata = OpenIDProviderMetadata(oidc.issuer)
+
+        # Further validation using token introspection
+        data = {"token": token, "token_type_hint": "access_token", "scope": "openid"}
+        auth = HTTPBasicAuth(oidc.client_id, oidc.client_secret)
+        response = requests.post(
+            oidc_metadata.introspection_endpoint, data=data, auth=auth
+        )
+        if response.status_code == 401:
+            logger.info("User's access token is invalid")
+            return None
+
+        introspection_resp = response.json()
+
+        if not introspection_resp["active"]:
+            logger.info("User is not active")
+            return None
+
+        if not any(
+            [
+                True
+                for role, value in introspection_resp[
+                    f"urn:zitadel:iam:org:project:{oidc.project_id}:roles"
+                ].items()
+                if oidc.organization_id in value
+            ]
+        ):
+            logger.info("User does not have roles in this organization")
+            return None
```

Comment:
I think the logic might be slightly wrong here. The authorizations (roles) are per project, I'm actually not entirely sure why the organization is returned here. Either way, I don't think we actually want to check that the organization ID matches, this works when the user is in the same organization, but might break if we ever do brokering etc in the future.

Reply:
So, to be clear, we really just want to check for the presence of a role in the configured project ID

Reply:
Should it check for specifically `OrgAdmin`/`User` roles or just any string at all?

Reply:
Lets check for OrgAdmin & User to be sure.